src/UnitTests/Scanning/VarargsFormatScannerTests.cs

```diff
@@ -0,0 +1,213 @@
+            this.vafs = CreateVaScanner(program);
+        }
+
+        private void Given_StackString(int stackOffset, string str)
+        {
+            uint uiAddr = 0x13;
+            var sp = m.Register(program.Architecture.StackRegister);
+            var stackAccess = m.IAdd(sp, stackOffset);
+            state.SetValueEa(stackAccess, Constant.Word32(uiAddr));
+            WriteString(program, uiAddr, str);
+        }
+
+        private void Given_RegString(string reg, string str)
+        {
+            uint uiAddr = 0x13;
+            var r = program.Architecture.GetRegister(reg);
+            state.SetRegister(r, Constant.Word32(uiAddr));
+            WriteString(program, uiAddr, str);
+        }
+
+        [Test]
+        public void Vafs_NoVarargs()
+        {
+            Given_VaScanner(x86);
+            var emptyChr = new ProcedureCharacteristics();
+            var emptySig = new FunctionType();
+            Assert.IsFalse(vafs.TryScan(null, null));
+            Assert.IsFalse(vafs.TryScan(emptySig, null));
+            Assert.IsFalse(vafs.TryScan(null, emptyChr));
+            Assert.IsFalse(vafs.TryScan(emptySig, emptyChr));
+            Assert.IsFalse(vafs.TryScan(x86PrintfSig, null));
+            Assert.IsFalse(vafs.TryScan(x86PrintfSig, emptyChr));
+            Assert.IsFalse(vafs.TryScan(null, printfChr));
+            Assert.IsFalse(vafs.TryScan(emptySig, printfChr));
+        }
+
+        [Test]
+        public void Vafs_X86Printf()
+        {
+            Given_VaScanner(x86);
+            Given_StackString(4, "%d %f");
+            Assert.IsTrue(vafs.TryScan(x86PrintfSig, printfChr));
+            var c = Constant.Word32(666);
+            var instr = vafs.BuildInstruction(c, new CallSite(4, 0));
+            Assert.AreEqual(
+                "0x0000029A(Mem0[esp:(ptr char)], Mem0[esp + 4:int32], " +
+                           "Mem0[esp + 8:real64])",
+                instr.ToString());
+        }
+
+        [Test]
+        public void Vafs_X86Sprintf()
+        {
+            Given_VaScanner(x86);
+            Given_StackString(8, "%c");
+            Assert.IsTrue(vafs.TryScan(x86SprintfSig, printfChr));
+            var ep = new ExternalProcedure("sprintf", x86SprintfSig);
+            var pc = new ProcedureConstant(new CodeType(), ep);
+            var instr = vafs.BuildInstruction(pc, new CallSite(4, 0));
+            Assert.AreEqual(
+                "sprintf(Mem0[esp:(ptr char)], Mem0[esp + 4:(ptr char)], " +
+                        "Mem0[esp + 8:char])",
+                instr.ToString());
+            var appl = (Application)((SideEffect)instr).Expression;
+            var sig = ((ProcedureConstant)appl.Procedure).Procedure.Signature;
+            Assert.AreEqual(
+                "(fn void ((ptr char), (ptr char), char))",
+                sig.ToString());
+        }
+
+        [Test]
+        [Ignore("Varargs scanning has not implemented on PowerPc")]
+        public void Vafs_PpcPrintf()
```

Comment:
I can take a look at this later. The calling convention typically is dictated primarily by the `Platform`, since in theory (and certainly in practice) an assembler language programmer can pass arguments any way she wants.